File: day_01/solution.java
import java.io.*;
import java.util.*;
import java.nio.file.*;
import java.util.stream.Collectors;

public class solution {
  public static void main(String[] args) throws IOException {
    List<List<Integer>> inputList = Files.lines(Paths.get("data.txt"))
      .map(line -> Arrays.stream(line.trim().split("\\s+"))
      .map(Integer::parseInt)
      .collect(Collectors.toList()))
      .collect(Collectors.toList());

    List<Integer> leftList = new ArrayList<>();
    List<Integer> rightList = new ArrayList<>();

    for (List<Integer> pair : inputList) {
      leftList.add(pair.get(0));
      rightList.add(pair.get(1));
    }

    // Part 1
    Collections.sort(leftList);
    Collections.sort(rightList);

    int distance = 0;
    for (int i = 0; i < leftList.size(); i++) {
      distance += Math.abs(leftList.get(i) - rightList.get(i));
    }

    System.out.println(distance);

    // Part 2
    Map<Integer, Long> rightCounter = rightList.stream()
      .collect(Collectors.groupingBy(e -> e, Collectors.counting()));

    long similarityScore = 0;
    for (int leftNumber : leftList) {
      similarityScore += leftNumber * rightCounter.getOrDefault(leftNumber, 0L);
    }

    System.out.println(similarityScore);
  }
}

File: day_01/solution.py
from collections import Counter


with open('1/data.txt') as f:
    input_list = [list(map(int, line.strip().split())) for line in f]

left_list = []
right_list = []
for i in range(len(input_list)):
    left_list.append(input_list[i][0])
    right_list.append(input_list[i][1])

# Part 1
left_list.sort()
right_list.sort()

distance = sum(abs(l - r) for l, r in zip(left_list, right_list))

print(distance)

# Part 2

right_counter = Counter(right_list)
similarity_score = sum(
    left_number * right_counter[left_number] for left_number in left_list)

print(similarity_score)
